File: src/07.greedy/02.subArrayGreedy/10.503.ts
export function nextGreaterElements(nums: number[]): number[] {
    const n = nums.length;
    const ans = new Array(n).fill(-1);
    const stack = [];

    for (let i = 0; i < n * 2; i++) {
        while (stack.length && nums[stack[stack.length - 1]] < nums[i % n]) {
            ans[stack.pop()] = nums[i % n];
        }
        stack.push(i % n);
    }

    return ans;
}

File: src/07.greedy/02.subArrayGreedy/10.503.go
package sub_array_greedy

func NextGreaterElements(nums []int) []int {
	n := len(nums)
	ans := make([]int, n)
	stack := []int{}

	for i := range ans {
		ans[i] = -1
	}

	for i := 0; i < n*2-1; i++ {
		idx := i % n
		for len(stack) > 0 && nums[stack[len(stack)-1]] < nums[idx] {
			ans[stack[len(stack)-1]] = nums[idx]
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, idx)
	}

	return ans
}
